adminapi/base.py

```diff
@@ -0,0 +1,450 @@
+        if hasattr(other, 'items'):
+            other = other.items()
+        for key, value in chain(other, kwargs.items()):
+            self[key] = value
+
+
+class MultiAttr(set):
+    """This class must redefine all mutable methods of the set class
+    to maintain the old values on the BaseServerObject.
+    """
+
+    def __init__(self, other, server_object, attribute_id, datatype=None):
+        super(MultiAttr, self).__init__(other)
+        self._server_object = server_object
+        self._attribute_id = attribute_id
+        self.datatype = self._validate(attribute_id, datatype)
+
+    def _validate(self, attribute_id, datatype):
+        """This method accepts class properties as arguments to be called
+        with normal sets."""
+        for elem in self:
+            datatype = _validate_value(attribute_id, elem, datatype)
+        return datatype
+
+    def __str__(self):
+        return ' '.join(str(x) for x in self)
+
+    def copy(self):
+        return MultiAttr(
+            self, self._server_object, self._attribute_id, self.datatype
+        )
+
+    def add(self, elem):
+        if elem in self:
+            return
+        self.datatype = _validate_value(
+            self._attribute_id, elem, self.datatype
+        )
+        self._server_object._save_old_value(self._attribute_id)
+        super(MultiAttr, self).add(elem)
+
+    def remove(self, elem):
+        if elem in self:
+            self._server_object._save_old_value(self._attribute_id)
+        super(MultiAttr, self).remove(elem)
+        if not self:
+            self.datatype = None
+
+    def discard(self, elem):
+        if elem in self:
+            self.remove(elem)
+
+    def pop(self):
+        for elem in self:
+            break
+        else:
+            raise KeyError()
+        self.remove(elem)
+        return elem
+
+    def clear(self):
+        if not self:
+            return
+        self._server_object._save_old_value(self._attribute_id)
+        super(MultiAttr, self).clear()
+        self.datatype = None
+
+    def update(self, *others):
+        if not others:
+            return
+        new = others[0]
+        for other in others[1:]:
+            new |= other
+        if not new - self:
+            return
+        MultiAttr._validate(new, self._attribute_id, self.datatype)
+        self._server_object._save_old_value(self._attribute_id)
+        super(MultiAttr, self).difference_update(new)
+
+    def intersection_update(self, *others):
+        if not others:
+            return
+        new = others[0]
+        for other in others[1:]:
+            new &= other
+        if not self - new:
+            return
+        self._server_object._save_old_value(self._attribute_id)
+        super(MultiAttr, self).intersection_update(new)
+
+    def difference_update(self, *others):
+        if not others:
+            return
+        new = others[0]
+        for other in others[1:]:
+            new |= other
+        if not self & new:
```

Comment:
Maybe store the result first
``` python
new &= self
```
or the like, as I misread it first like
``` python
if not self and new:
    ...
```

btw. wouldn't it be the same here?

Reply:
This might even be wrong.  I am too tired to lazy to analyse this again.  I don't think anybody is using them anyway.  We gain so little by checking if the operation would change the values.  I will make them all just save the value and call the implementation.  We can later check later if the old values and new values are the same.  This would even cover the case when you set an attribute to some value and then set it back to its existing value.

Reply:
Agree, gain would be too less to investigate if further.